### PR TITLE
[FIRRTL] Canonicalize reductions looking through casts of various forms

### PR DESCRIPTION
**Pull request by darthscsi**
Greatly reduces the number of ops in real designs
```
wc -l foo.mlir foo5.mlir 
  2247057 foo.mlir
  2057784 foo5.mlir
```

### CODE REVIEW

**Comment by darthscsi** [comment]
TODO: tests

File: include/circt/Dialect/FIRRTL/FIRRTLExpressions.td [hunk 1]
```diff
@@ -674,26 +674,28 @@ let hasCanonicalizer = true in
...
-      returns 0 for zero-bit-wide operands.
-    }];
-  }
-  def XorRPrimOp : UnaryPrimOp<"xorr", IntType, UInt1Type> {
-    let description = [{
-      Horizontally reduce a value to one bit, using the 'xor' operation to merge
-      bits.  `xorr(x)` is equivalent to `popcount(concat(x, 1b0)) & 1`.  As
-      such, it returns 0 for zero-bit-wide operands.
-    }];
+  let hasCanonicalizer =1 in {
```
Comment (fabianschuiki) [inline include/circt/Dialect/FIRRTL/FIRRTLExpressions.td:677]:
```suggestion
  let hasCanonicalizer = 1 in {
```

**Review by fabianschuiki — APPROVED** [review]
LGTM. Juicy reduction 😎

File: include/circt/Dialect/FIRRTL/FIRRTLCanonicalization.td [hunk 2]
```diff
@@ -326,6 +336,80 @@ def XorOfPad : Pat <
...
+  (MoveNameHint $old, (OrRPrimOp $x)),
+  [(KnownWidth $x), (IntTypes $x)]>;
+
+/// orr(asUInt(x)) -> orr(x)
+def OrRasUInt : Pat <
+  (OrRPrimOp:$old (AsUIntPrimOp $x)),
+  (MoveNameHint $old, (OrRPrimOp $x)),
+  [(KnownWidth $x), (IntTypes $x)]>;
+
+/// orr(cvt(x)) -> orr(x)
```
Comment (dtzSiFive) [inline include/circt/Dialect/FIRRTL/FIRRTLCanonicalization.td:351]: Please add tests for these, at least somewhat?

I'll go through these more carefully soon, but would be good to see them in action and ensure they're working as expected.

**Review by dtzSiFive — COMMENTED** [review]
Awesome, thanks for adding all these, and hooray for the reduction!!

Will revisit (don't mean to block necessarily but folding logic is something usually needs very careful inspection/review) but great direction and look forward to seeing this landed.

**Comment by darthscsi** [comment]
> usually needs very careful inspection

Indeed it does.

File: include/circt/Dialect/FIRRTL/FIRRTLCanonicalization.td [hunk 3]
```diff
@@ -282,6 +282,16 @@ def AndOfPad : Pat <
...
 
+/// cvt is adding a 0, which will pass through the and
+/// and(cvt(x:U), const) -> cat(0, and(x,trunc(const))
+def AndCvtU : Pat <
+  (AndPrimOp:$old (CvtPrimOp $x), (ConstantOp:$ocst $cst)),
+  (MoveNameHint $old, (CatPrimOp 
+    (NativeCodeCall<"$_builder.create<ConstantOp>($0.getLoc(), APSInt(APInt()))"> $ocst),
+    (AndPrimOp $x, (NativeCodeCall<"$_builder.create<ConstantOp>($0.getLoc(), cast<IntType>($0.getType()), $1.getValue().trunc($1.getValue().getBitWidth()-1).zext(*cast<IntType>($0.getType()).getWidth()))"> $x, $cst))
+  )),
+  [(KnownWidth $x), (UIntType $x)]>;
```
Comment (seldridge) [inline include/circt/Dialect/FIRRTL/FIRRTLCanonicalization.td:293]: There is a variant of this pattern that applies for all of `OrCvtU`, `XorCvtU`, and `NotCvtU`. (Anytime there is a `cvt` whose user is an op which does an `SInt` -> `UInt` conversion.) The patterns should be the following:

- `or(cvt(x:U), const) -> cat(head(const, 1), or(x, trunc(const))`
- `xor(cvt(x:U), const) -> cat(head(const, 1), xor(x, trunc(const)))`
- `not(cvt(x:U)) -> cat(1, not(x))`

Did you experiment with these, too?

Comment (darthscsi) [reply include/circt/Dialect/FIRRTL/FIRRTLCanonicalization.td:293]: I didn't see them in the code I was looking at.  There were a lot of add/sum after a cvt which might be able to be narrowed.

Then would be a good thing to add at some point.

**Pull request merged**